Mean power detection

MichaelCotton opened the pull request:
I cleaned some stuff up and added a `Detection` object to support lab y factor cals. @djanderson gave it a first review. @linhvu1 and @bradeales : Pls give it a look.  I hope this gets @linhvu1 one step closer to using scos-sensor.

MichaelCotton committed:
Changing measurement types to support lab cals

- Added `Detection` object that describes algorithm applied to IQ samples
- Reduced `SingleFrequencyFFTDetection` to `FFTDetection`to describe algorithm applied to DFT(IQ samples)
- Added more specificity to detector names in `SingleFrequencyDetection`, e.g., "fft_mean_power" instead of "mean_power"
- Generalized `SteppedFrequencyDetection` to support range of single freq detectors - add "frequencies" array
- Provided simple example for `SteppedFrequencyDetection`
- Changed array name to "results"

MichaelCotton committed:
Typo

MichaelCotton committed:
Typos2

- reordered elements of `FFTDetection`
- fixed description of `SteppedFrequencyDetection`

MichaelCotton committed:
Resolve issue #6

Provided some specifics to arrays

MichaelCotton committed:
Resolve issue #4

MichaelCotton committed:
Resolve issue #3

- Added specifics to `horizontal_gain_pattern` and `vertical_gain_pattern`

MichaelCotton committed:
Removed `results`

bradeales commented on README.md:
```diff
@@ -197,28 +198,51 @@ Per SigMF, the annotations value is an array of annotation segment objects that
 
-##### SingleFrequencyFFTDetection Object
-Single-frequency FFT detection is a standard software-defined radio measurement. The `SingleFrequencyFFTDetection` object contains the following name/value pairs:  
+##### Detection Object
```
What is the purpose of this object? How does it differ from all the other detection objects? Do you mean it is a generic or basic detection algorithm? Maybe call it "BasicDetection Object" or something like that? As "Detection Object", it makes it seem it's a parent object to all the others.

MichaelCotton replied:
I have been struggling with this. The purpose is to provided metrics on the time series. Detection is a stretch, but I don't know what else to call it. Perhaps TimeSeriesCalculations?

MichaelCotton replied:
Instead of `Detection` v `FFTDetection`, `TimeDomainDetection` v `FrequencyDomainDetection`?

bradeales commented on README.md:
```diff
@@ -184,7 +185,7 @@ Per SigMF, the annotations value is an array of annotation segment objects that
 |`altitude`|false|float|meters|The height of the antenna above mean sea level.|
 |`environment`|false|string|N/A|A description of the environment where antenna is mounted. E.g. `"indoor"` or `"outdoor"`.|
-|`measurement_type`|true|object|N/A|The type of measurement acquired: [SingleFrequencyFFTDetection](#singlefrequencyfftdetection-object), [SteppedFrequencyFFTDetection](#steppedfrequencyfftdetection-object), [SweptTunedDetection](#swepttuneddetection-object) or [YFactorCalibration](#yfactorcalibration-object).|
+|`measurement_type`|true|object|N/A|The type of measurement acquired, e.g., [SteppedFrequencyDetection](#steppedfrequencydetection-object), [SweptTunedDetection](#swepttuneddetection-object) or [YFactorCalibration](#yfactorcalibration-object).|
```
Do you want to list all the new detection types here as well?

MichaelCotton replied:
The new detection types are a subsets of `SteppedFrequencyDetection`. Basically, if you specify one frequency then it reduces to new detection types.

bradeales commented on README.md:
```diff
@@ -104,8 +105,8 @@ The `Antenna` object contains the following name/value pairs:
 |`gain`|false|float|dBi|Antenna gain in direction of maximum radiation or reception.|
-|`horizontal_gain_pattern`|false|array|dBi|Antenna gain pattern in horizontal plane.|
-|`vertical_gain_pattern`|false|array|dBi|Antenna gain pattern in vertical plane.|
+|`horizontal_gain_pattern`|false|array of floats|dBi|Antenna gain pattern in horizontal plane from 0 to 359 degrees in 1 degree steps.|
```
`float` is not needed for 1 degree steps as per description. "array of integers" perhaps?

bradeales replied:
Also, why do we need an array, is it not just a list?

MichaelCotton replied:
Floats are needed for antenna gain numbers. The 0 to 359 count describe the azimuth angles to which the gains correspond.

bradeales commented on README.md:
```diff
@@ -104,8 +105,8 @@ The `Antenna` object contains the following name/value pairs:
-|`horizontal_gain_pattern`|false|array|dBi|Antenna gain pattern in horizontal plane.|
-|`vertical_gain_pattern`|false|array|dBi|Antenna gain pattern in vertical plane.|
+|`horizontal_gain_pattern`|false|array of floats|dBi|Antenna gain pattern in horizontal plane from 0 to 359 degrees in 1 degree steps.|
+|`vertical_gain_pattern`|false|array of floats|dBi|Antenna gain pattern in vertical plane from -90 to +90 degrees in 1 degree steps.|
```
Do we need an array if we specify in the description the start, end, and increment? List?

bradeales reviewed (changes requested):
Biggest question is about new "Detection Object".

MichaelCotton committed:
Resolve Detection and FFT Detection naming issue

- Adopt TimeDomainDetection and FrequencyDomainDetection

MichaelCotton committed:
Distinguish measurement from detection

MichaelCotton committed:
Removed Control Plane section

- Moved control plane text to description of `ScheduleEntry` object
- Removed Control Plane and Data Plane sections
- Moved rest up one layer
- Updated Description text

MichaelCotton committed:
Established TransmitterDefinition object

- Reordered things to make it more readable - sorry this probably annoys people
- Changed object name from `SystemToDetect` to `TransmitterDefinition`, which is more intuitive to support propagation measurements which requires a known transmitter
- Utilized the `Antenna` object in `TransmitterDefinition` object
- Utilized the sigmf-ns-waveform namespace in `TransmitterDefinition` object

MichaelCotton committed:
typo

MichaelCotton committed:
added transmitter_identification

- Moved `transmitter_definition` to global to support measurements of known transmitters, e.g., prop measurements, occupancy measurements of targeted signals, that will not change of the course of a capture
- Added `transmitter_identification` to annotations to support transmitter signal identification and localization
- `TransmitterDefinition` object has link to sigmf-ns-waveform

bradeales commented:
I like it :+1:. Every question I could think of seems to be covered!

djanderson commented on README.md:
```diff
@@ -146,20 +128,22 @@ Each `RFPath` object contains the following name/value pairs:
+|`altitude`|false|float|meters|Altitude above mean sea level.|
+
+### 3.3 ScheduleEntry Object
+The `ScheduleEntry` object is associated with the SCOS control plane, which is implemented through the use of a RESTful API residing on the sensor, see [SCOS Sensor](https://github.com/NTIA/scos-sensor) and [SCOS Control Plane API Reference](https://ntia.github.io/scos-sensor/). A sensor advertises its **capabilities**, among which are **actions** that users can schedule the sensor to do. Sensor  ...
```
"are schedule to" -> "are scheduled to" in the last sentence.

djanderson commented on README.md:
````diff
@@ -174,54 +158,116 @@ The `ScheduleEntry` object contains the following name/value pairs:
+    "reference": "output of isotropic antenna"
+  }
+}
+```
````
1) This isn't valid JSON (missing comma after frequencies array, missing colon after algorithm). I recommend dropping any JSON examples into an online JSON validator/formatter just to make sure I didn't miss anything as well. 

2) I think this is trying to be just a bit too concise to the point of making it difficult to interpret. Remember that when you use a `TimeDomainDetection` object or a `FrequencyDomainDetection` object, they're not actually labeled as such. So I'm looking at this and thinking "okay so to figure out if this is a time domain or frequency domain stepped measurement, I'm going to need to look at the different fields in both objects, pick a field that's required in one but doesn't exist in the other and look for that"... etc.

Quick off-the-cuff suggestion: `detection_domain: "time"|"frequency"`?

MichaelCotton committed:
Response to DA requested changes

Good suggestions...
- "are schedule to" -> "are scheduled to" in the last sentence of `ScheduleEntry` object
- json examples fixed and validated
- added `detection_domain` to ```TimeDomainDetection``` and ```FrequencyDomainDetection``` objects
